File: layrz_sdk/entities/asset.py
"""Asset Entity"""

import sys
from typing import Any, Optional

from pydantic import BaseModel, Field, model_validator

if sys.version_info >= (3, 11):
  from typing import Self
else:
  from typing_extensions import Self

from .asset_operation_mode import AssetOperationMode
from .custom_field import CustomField
from .device import Device
from .sensor import Sensor


class Asset(BaseModel):
  """Asset entity definition"""

  pk: int = Field(description='Defines the primary key of the asset')
  name: str = Field(description='Defines the name of the asset')
  vin: Optional[str] = Field(
    default=None,
    description='Defines the serial number of the asset, may be an VIN, or any other unique identifier',
  )
  plate: Optional[str] = Field(default=None, description='Defines the plate number of the asset')
  asset_type: Optional[int] = Field(description='Defines the type of the asset', alias='kind_id', default=None)
  operation_mode: AssetOperationMode = Field(description='Defines the operation mode of the asset')
  sensors: list[Sensor] = Field(default_factory=list, description='Defines the list of sensors of the asset')
  custom_fields: list[CustomField] = Field(
    default_factory=list, description='Defines the list of custom fields of the asset'
  )
  devices: list[Device] = Field(default_factory=list, description='Defines the list of devices of the asset')
  children: list[Self] = Field(default_factory=list, description='Defines the list of children of the asset')

  @model_validator(mode='before')
  def _validate_model(cls: Self, data: dict[str, Any]) -> dict[str, Any]:
    """Validate model"""
    operation_mode: Optional[str] = data.get('operation_mode')
    if operation_mode == AssetOperationMode.ASSETMULTIPLE.name:
      data['devices'] = []

    else:
      data['children'] = []

    return data

  @property
  def primary(self: Self) -> Optional[Device]:
    """Get primary device"""
    if self.operation_mode not in [AssetOperationMode.SINGLE, AssetOperationMode.MULTIPLE]:
      return None

    for device in self.devices:
      if device.is_primary:
        return device

    return None

File: go/entities/asset.go
package entities

// Asset represents an asset entity.
type Asset struct {
	// Id is the unique identifier of the asset.
	Id int64 `json:"id"`
	// Name of the asset
	Name string `json:"name"`
	// Optional Vehicle Identification Number or any other unique identifier
	VIN *string `json:"vin,omitempty"`
	// Optional License plate number of the asset, or any other secondary identifier
	Plate *string `json:"plate,omitempty"`
	// Optional KindId representing the type or category of the asset
	KindId *int64 `json:"kind_id,omitempty"`
	// OperationMode indicates the current operation mode of the asset
	OperationMode *AssetOperationMode `json:"operation_mode,omitempty"`
	// StaticPosition is the fixed position of the asset when in STATIC mode
	StaticPosition *StaticPosition `json:"static_position,omitempty"`
	// Points to the primary device associated with the asset
	Points []StaticPosition `json:"points,omitempty"`
	// PrimaryId is the identifier of the primary device linked to the asset
	PrimaryId *int64 `json:"primary_id,omitempty"`
	// Contacts associated with the asset
	Contacts []AssetContact `json:"contacts,omitempty"`
	// OwnerId is the identifier of the user who owns the asset
	OwnerId *int64 `json:"owner_id,omitempty"`
	// PartitionNumber is the partition number the asset belongs to, useful on multi-tenant systems
	PartitionNumber *int64 `json:"partition_number,omitempty"`
	// Devices associated with the asset, only populated on Operation Mode is not ASSETMULTIPLE
	Devices []Device `json:"devices,omitempty"`
	// Children assets, only populated when Operation Mode is ASSETMULTIPLE
	Children []Asset `json:"children,omitempty"`
	// CustomFields holds additional custom data for the asset
	CustomFields []CustomField `json:"custom_fields,omitempty"`
	// Sensors associated with the asset
	Sensors []Sensor `json:"sensors,omitempty"`
}

// Primary returns the Device of the primary device associated with the asset
func (a *Asset) Primary() *Device {
	return nil
}

// AssetType returns the KindId of the asset, it's an alias for better readability
func (a *Asset) AssetType() *int64 {
	return a.KindId
}
